packages/backend/src/core/events/generateEventRecords.ts

```diff
@@ -0,0 +1,70 @@
+import { UnixTime } from '@l2beat/common'
+
+import { EventRecord } from '../../peripherals/database/EventRepository'
+import { EventDetails } from './types/EventDetails'
+
+export function generateEventRecords(
+  { name, projectId }: EventDetails,
+  logs: bigint[],
+  timestamps: { timestamp: UnixTime; blockNumber: bigint }[],
+): EventRecord[] {
+  if (timestamps.length === 0) {
+    return []
+  }
+  if (timestamps[0]?.timestamp.toNumber() % 86400 !== 3600) {
+    throw new Error('Algorithm works only if first timestamp is 01:00')
+  }
+
+  const sortedLogs = logs.sort((a, b) => Number(a - b))
+
+  let i = 0
+  const hourly: EventRecord[] = []
+  const sixHourly: EventRecord[] = []
+  const daily: EventRecord[] = []
+
+  for (const { timestamp, blockNumber } of timestamps) {
+    let count = 0
```

Comment:
Why not have `hourlyCount`, `sixHourlyCount` and `dailyCount` and just reset them to zero, avoiding the need to filter and reduce.